### ISSUE TITLE
High-resolution Bitmap replacement

### ISSUE THREAD

**Issue by Splendide-Imaginarius** [mkxp-z/mkxp-z#15]
RGSS typically runs at 640x480; it would be highly useful to facilitate running at higher resolutions. I have an interesting approach to this: high-resolution Bitmap replacement.

**Background**

If you're familiar with N64 and GameCube emulators, you've probably seen their high-resolution texture replacement feature. Conceptually, that's similar to what I'm proposing, but the implementation is pretty different, because mkxp-z controls much of the engine and can therefore handle this use case much more intelligently than emulators can. That said, from an end-user's point of view, it looks pretty similar: put replacement images in a designated folder, and they show up in the game.

**Is this useful to game devs?**

Yes. It's fairly common for RPG Maker game devs to ship high-resolution versions of the game's artwork along with the game (1600x1200 is a common resolution). I'm pretty sure that the devs would include this artwork in the game itself if the engine supported it.

**Is this useful to game modders?**

Yes. N64 and GameCube graphical modding communities are quite active; I don't see any reason why RGSS games couldn't spring a similar community. There are three main ways that modding could be done:

1. As mentioned above, existing games often ship with 1200p artwork; inserting these into the game itself is an easy modding approach.
2. AI-based upscalers like Real-ESRGAN do a pretty good job on a lot of RGSS game assets; these tools are commonly used by modders.
3. Manual re-touching or replacement is also a thing (and is pretty common as well).

(For what it's worth, I'm personally intending to release at least one game mod using this, so it definitely won't go completely unused.)

**Why patch mkxp-z instead of patching the Ruby game code?**

I tried Ruby patches first, actually. It was extremely messy, and eventually I ran into a brick wall because mkxp-z makes a bunch of assumptions about tileset size. It's theoretically possible to upscale the CG artwork (and I got this mostly working, though not without glitches galore), but increasing the rendering resolution results in maps being zoomed out rather than simply making them look sharper. Implementing this in mkxp-z itself took much less time, produced far fewer glitches, doesn't introduce conflicts with the tile size, and should work well with arbitrary games (the Ruby-side patches I made were game-specific, which doesn't scale as well effort-wise).

**What does the code look like?**

There are two main things I had to do with mkxp-z's codebase.

1. The Bitmap class was patched to store an extra Bitmap object with a larger width/height, and the methods of the class were patched to apply the relevant operations to the extra Bitmap as well (with coordinate scaling as appropriate).
2. The Graphics class was patched to use a higher-resolution framebuffer, so that the higher-resolution Bitmaps don't get scaled-down when rendered to the screen, and the functions that render to the screen framebuffer were patched to scale the transformation matrix as needed.

All things considered, it wasn't too hard, and the patches don't seem too invasive. Took me a few weekends to fully implement, and at this point I just need to clean up some debug leftovers and gate the entire thing behind a config option.

**What images can be high-res?**

CG images and rendered text work fine. Sprites also appear to work fine, though I didn't test them as thoroughly. Tilesets can't be high-res yet, because the Atlas texture uses different code from everything else. Probably I could make Tilesets work, but I didn't try yet. That said, Tilesets don't conflict with sprites/CG's, e.g. you can have high-res sprites/CG's on the screen while standard-res Tilemaps are in use, and everything looks fine (it doesn't zoom out the map or cause any other breakage).

**Does it work with the full Bitmap API?**

A few of the more unusual Bitmap features (e.g. animations) aren't supported by my high-res patches. It shouldn't be hard to add support for the missing features, if someone really wants to play games in high-res that use those features.  I've only tested so far with one game, and I implemented all the Bitmap API features that that game used.

**In conclusion**

I think this would be a pretty cool feature to add to mkxp-z. mkxp-z already has some nice graphical features that standard RPG Maker doesn't have (e.g. unlocked screen resolution, modified interpolation algorithms, etc.); this seems like a good addition to that repertoire. I'm not asking anyone to implement it (as mentioned, I have it basically working) -- I'm basically just wondering if such a PR would be welcomed, and whether there are any particular requirements/constraints I need to take into account to make such a PR as easy to review as possible.

Thanks for reading!

**Comment by hcphoenix** [comment]
That sounds really cool if you got it working!

Based on discussion that happened on Discord you might be better off making a fork than a pull request, since mkxp-z is intended to be feature-complete at this point. That said, I'm really excited to try it!

**Comment by Splendide-Imaginarius** [comment]
@hcphoenix I'm hesitant to take on the responsibility of maintaining a long-term mkxp fork, especially since I don't have a strong understanding of the codebase or a lot of the underlying libraries. (I'm quick enough at learning that I can get the changes I want to work, but it's not a substitute for the high-quality codebase familiarity that @strumajen has.) If it's absolutely necessary, I could maybe be persuaded to maintain a fork, but it's really something I'd only do as a last resort if neither @strumajen nor anyone else wants to maintain the codebase. (I was hoping to hear some clarification from @strumajen before I worry about this -- and I'm OK with waiting a bit until they have some time to answer.)

That said, I'm curious -- was this a Discord discussion specifically about this GitHub issue, or something more general? (Unfortunately I don't use Discord at the moment.) Also, may I ask whether you're interested in trying it as a game developer or a game modder?

**Comment by hcphoenix** [comment]
> (I was hoping to hear some clarification from @strumajen before I worry about this -- and I'm OK with waiting a bit until they have some time to answer.)

Apologies, I feel pretty awkward speaking for anyone, which is why I waited a couple days after people had stopped discussing this on Discord before getting strumajen's permission to write something here. Strumajen wrote that "maybe original mkxp can have it [this feature]," which I can't really speak to at all but might be worth trying if you're looking for another maintainer.

> Also, may I ask whether you're interested in trying it as a game developer or a game modder?

My interest would be as a modder already actively working on updating an old RMXP game.

**Comment by Splendide-Imaginarius** [comment]
@hcphoenix Alright, that's useful info to have. I'll ponder it a bit. Should I infer that there's actually nontrivial interest on Discord in this feature?

**Comment by hcphoenix** [comment]
Well, we're only talking about 4 or 5 people who were discussing this, but although they were impressed you got it working I'm not aware of anyone else with specific plans to use this feature (except possibly for better font rendering?). The primary userbase for mkxp-z as it currently exists seems to be people porting or developing Pokemon Essentials games or other games using pixel art where nearest neighbor scaling is sufficient. I really hope that doesn't stop you from sharing your work in some form though! Even if my reasons are a little selfish it'd be cool to see other people helping to grow this very specific niche. 😅

**Comment by Splendide-Imaginarius** [comment]
> Well, we're only talking about 4 or 5 people who were discussing this, but although they were impressed you got it working I'm not aware of anyone else with specific plans to use this feature

@hcphoenix Honestly I admit I'm kind of surprised there's literally any interest at this point. I was kind of expecting that I'd release the code, and then release an example game mod to showcase the capabilities, and then post about it on a few forums or whatever, and only then attract any interest. So, even one person with plans to use it, before I even pushed the code, is certainly a pleasant surprise.

> (except possibly for better font rendering?)

Ironically that's the component of my patches that looks the least impressive. Mainly because a lot of RPG Maker games seem to use fonts that, for whatever reason, look a tad weird when rendered at large font sizes (they don't look jagged, just a little weird). I imagine that's fixable with font substitution (which mkxp-z already has) but I haven't tried it yet. That said, sure, if people want to use my patches for that, I'm not going to complain.

> using pixel art where nearest neighbor scaling is sufficient

Heh, this is very much a subjective thing. Some of my friends like nearest-neighbor a lot, personally I find it distracting to the point that the game is nearly unplayable. Humans are diverse, I guess.

> I really hope that doesn't stop you from sharing your work in some form though!

It won't stop me from sharing it; I wrote this code because I was curious to see if I could do it (not because I was betting on getting a lot of users), and I'm generally not in the habit of keeping working code to myself. There *will* be potential delays in terms of how much time I can spend on this (it's purely a hobby and I have a real job), and that will be exacerbated if I wind up having to maintain a fork myself, but hopefully I'll be able to sort that out.

> Even if my reasons are a little selfish it'd be cool to see other people helping to grow this very specific niche.

Doesn't seem selfish to me FWIW - I'm just pleasantly surprised that anyone wants to use it before I even made a proper announcement. Sometimes niche projects are the coolest ones. :)

One quick question, if you don't mind. Are there any actively maintained mkxp forks besides mkxp-z that you're aware of? I found a couple of them on GitHub with a search yesterday (which I wasn't aware of previously). Especially if you know of any that are maybe more likely to be amenable to merging this kind of thing, I'd like to reach out to them before I worry about the possibility of having to maintain my own fork.

**Comment by hcphoenix** [comment]
Good news! Strumajen said they'll probably take the PR if it fixes text at arbitrary window sizes.

**Comment by Splendide-Imaginarius** [comment]
> Good news! Strumajen said they'll probably take the PR if it fixes text at arbitrary window sizes.

@hcphoenix Oh cool! What's the definition of "fixes" here? As best I can tell, text looks fine in mkxp-z with arbitrary window sizes already, except that text gets blurry as you scale it (at least, that's what I observe with bilinear filtering turned on). My patches do fix the blurriness of scaled text in such cases, but I'm not sure if there's something else I've failed to notice about the status quo that needs fixing.

EDIT: If desired, I could probably upload a screenshot of text with and without the patch. Let me know if this would be helpful.

Also, quick status update: my patches originally had an artificial restriction that the scaling factor had to be an integer. I had this restriction there due to concerns that non-integer scaling factors might introduce rendering bugs, and I didn't want to deal with such bugs until the rest of the patches were stable. Today I decided to try removing that restriction, and... everything still works fine. Tested both 1.5x scaling and 2.5x scaling, no apparent glitches. So, whatever I end up releasing will support non-integer scaling factors. I also did some minor cleanup of the patches (e.g. removing a few stray log statements and commented-out code lines that were leftover from testing). Definitely getting closer to a state where I'd be comfortable publishing the code.

**Comment by Splendide-Imaginarius** [comment]
> Tilesets can't be high-res yet, because the Atlas texture uses different code from everything else. Probably I could make Tilesets work, but I didn't try yet.

I got high-res Tilesets working today (at least for VX Ace). Inching closer to publishing the code; main thing remaining is gating everything behind a config option.
